nservicebus/messaging/evolving-contracts.md

```diff
@@ -1,108 +1,78 @@
+* Endpoints updated to the latest message contract might send messages using the new contract to endpoints that are still deployed with the old contract version.
+
+Generally the problem can't be resolved at the infrastructure level; therefore NServiceBus users must analyze their individual systems, consider how they are expected to evolve, and define the strategy which will make the most sense in their particular circumstances.
+
+This article presents basic guidelines for choosing contracts evolution strategy, avoiding common mistakes, and ensuring that contracts will be easy to evolve over time.
+
+TODO: link to new document
+Note: Ensure that message contracts follow the general [messages design guidelines](/nservicebus/messaging/messages-events-commands.md#designing-messages).
+
+## Adding new message contracts
+
+When adding new message contracts to a contracts assembly, ensure to update the endpoints receveiving or subscribing to the new contract type first. When updating the sender/publisher first, the messages fail to be processed by endpoints that are not using the new contract assembly.
+
+## Adding data to existing contracts
+
+Adding additional data to existing contracts is the most common change of contracts. There are differnt approaches available.
+
+### Create a new contract type
+
+1. Similar to adding a completely new contract using the original contracts name + some version postfix (e.g. `CreateOrderV2`), create a new message contract by copying the existing contract and adding the additional data to the copy.
+1. Add message handlers for the new contract type to the receiving endpoints. Keep the existing mesage handlers for the "old" contract.
+1. Update the sender/publisher to send messages using the new contract type.
+1. Once all endpoints have been updated, and no more messages using the old contract type is left in the queue, the old type and associated handlers can be safely removed.
+
+This approaches requires all receivers to be updated first, as they will be aware of the new contract type before upgrading which results in message processing failures.
+
+TODO: what about version namespaces?
+
+### Add new properties to existing contract types
+
+1. Add the new properties to the existing type.
+1. Update senders to the new contract version, setting the additional data on the message
+1. Update receivers to handle the new contract version.
+
+This approach requires less modifications to the contracts and even allows to update senders and receivers in arbitrary order as long as the receivers can handle the absence of a value for the new property.
+TODO: can serialzers handle the absence?  
+
+It is recommended to use nullable types for the new properties to allow receivers to identify whether it is dealing with an old version of the contract.

```

Comment:
I primarily took that statement from the original document without thinking too much about it tbh. Depending on your migration approach (receiver- or sender-first), it might be very useful or not necessary at all. Primitive types are tricky without nullables, e.g. I think we even had (or still have?) some limitations in core that saga correlation properties can't be `0` because in that case we don't know whether the sender didn't set it or whether the sender really wanted it to be 0. Nullables can solve this properly.